### DISCUSSION TITLE
API route body size limit and authenticated binary download

### DISCUSSION THREAD

**Discussion by Tetedeiench** [vercel/next.js#28512] [Help]
Hi everyone ! 

I run a website built upon NextJS, and it is used to serve binaries of my own app. The app comes into different flavors : some require authentication, some do not (think of a free version and paid editions). 

Until now, I was using an API route for downloading the app - it checked the requested binary, handled the authentication, and served the binary to the client if all conditions were met. Simple and efficient design, very elegant, worked really fine.

While developing, I stumbled upon this in the console : "API response for /*hidden* exceeds 4MB. This will cause the request to fail in a future version. https://nextjs.org/docs/messages/api-routes-body-size-limit"

I looked far and wide and no one talks about this. It's a killer for me, as this basically means an API route cannot serve a binary larger than 4MB back to the client for some reason.

I'm back to square one : I can't host the files in /public (some requires authentication before being downloadable), cannot use the suggested AWS/whatever hosting (the pricing is prohibitive given my trafic), can't use API pages anymore to do that.

Is there any alternative using nextJS to serve a binary to a client while checking authentication, or am I doomed to build something else to do so ?

I can't even proxy the query to my backend using api routes, as my backend response will be larger than 4MB as well
and I don't want my backend to answer directly to requests from the outside world for security reasons.

I'm using NextAuth for authentication, and rely on some simple, custom logic before serving a file back to the client.

Thanks for your help !

**Comment by Tetedeiench** [comment]
Additional info : I'm getting this warning when testing locally. My website is self-hosted ( docker container built upon node:alpine ) on my own dedicated server.

I am not using Vercel services - I'm thus not sure I'm concerned by this warning.

**Reply by crice88** [reply]
I am in the same boat and am having trouble finding a solution. My problem is similar though I am using streams to pipe the file contents into the response so I can get immediate feedback when I click download. Seems silly that this warning pops as I should be able to make these choices on my own. If it's not performant, that's on me.

**Comment by 2stash** [comment]
@Tetedeiench @crice88 Seems like this is an intended issue, and also seems like there is no intention of changing it. 

See recent opened and closed issue: https://github.com/vercel/next.js/issues/33162

Documentation for 4mb limit: https://nextjs.org/docs/messages/api-routes-body-size-limit

**Comment by Tetedeiench** [comment]
@2stash @crice88 This is intended, but badly implemented. It's a fix for Vercel / cloud services that is forcefully applied to everyone, even outside cloud services. 

If this will fail in the future, this will prevent valid use cases, such as gating downloads behind auth mechanism.

I'm going to react in the issue @2stash linked, and I invite you to do the same - it was closed way too soon without considering the gist of the issue IMHO.

**Reply by crice88** [reply]
@Tetedeiench I will do the same. I had a feeling that it had something to do with Vercel...

**Comment by davidfaj** [comment]
This was answered and closed by https://github.com/vercel/next.js/issues/33162
